esphome/components/gc9a01/gc9a01_display.cpp

```diff
@@ -0,0 +1,241 @@
+  int r = color_8bit >> 5;
+  int g = (color_8bit >> 2) & 0x07;
+  int b = color_8bit & 0x03;
+  uint16_t color = (r * 0x04) << 11;
+  color |= (g * 0x09) << 5;
+  color |= (b * 0x0A);
+
+  return color;
+}
+
+uint8_t GC9A01Display::convert_to_8bit_color_(uint16_t color_16bit) {
+  // convert 16bit color to 8 bit buffer
+  // uint8_t r = color_16bit >> 11;
+  // uint8_t g = (color_16bit >> 5) & 0x3F;
+  // uint8_t b = color_16bit & 0x1F;
+
+  // return ((b / 0x0A) | ((g / 0x09) << 2) | ((r / 0x04) << 5));
+  uint8_t b = color_16bit >> 11;
+  uint8_t g = (color_16bit >> 5) & 0x3F;
+  uint8_t r = color_16bit & 0x1F;
+
+  return ((b / 0x0A) | ((g / 0x09) << 2) | ((r / 0x04) << 5));
+}
+
+void GC9A01Display::fill(Color color) {
+  auto color565 = display::ColorUtil::color_to_565(color);
+  memset(this->buffer_, convert_to_8bit_color_(color565), this->get_buffer_length_());
+  this->x_low_ = 0;
+  this->y_low_ = 0;
+  this->x_high_ = this->get_width_internal() - 1;
+  this->y_high_ = this->get_height_internal() - 1;
+}
+
+void GC9A01Display::fill_internal_(Color color) {
+  this->set_addr_window_(0, 0, this->get_width_internal(), this->get_height_internal());
+  this->start_data_();
+
+  auto color565 = display::ColorUtil::color_to_565(color);
+  for (uint32_t i = 0; i < (this->get_width_internal()) * (this->get_height_internal()); i++) {
+    this->write_byte(color565 >> 8);
+    this->write_byte(color565);
+    buffer_[i] = 0;
+  }
+  this->end_data_();
+}
+
+void HOT GC9A01Display::draw_absolute_pixel_internal(int x, int y, Color color) {
+  if (x >= this->get_width_internal() || x < 0 || y >= this->get_height_internal() || y < 0)
+    return;
+
+  // low and high watermark may speed up drawing from buffer
+  this->x_low_ = (x < this->x_low_) ? x : this->x_low_;
+  this->y_low_ = (y < this->y_low_) ? y : this->y_low_;
+  this->x_high_ = (x > this->x_high_) ? x : this->x_high_;
+  this->y_high_ = (y > this->y_high_) ? y : this->y_high_;
+
+  uint32_t pos = (y * width_) + x;
+  auto color565 = display::ColorUtil::color_to_565(color);
+  buffer_[pos] = convert_to_8bit_color_(color565);
+}
+
+// should return the total size: return this->get_width_internal() * this->get_height_internal() * 2 // 16bit color
+// values per bit is huge
+uint32_t GC9A01Display::get_buffer_length_() { return this->get_width_internal() * this->get_height_internal(); }
+
+void GC9A01Display::start_command_() {
+  this->dc_pin_->digital_write(false);
+  this->enable();
+}
+
+void GC9A01Display::end_command_() { this->disable(); }
+void GC9A01Display::start_data_() {
+  this->dc_pin_->digital_write(true);
+  this->enable();
+}
+void GC9A01Display::end_data_() { this->disable(); }
+
+void GC9A01Display::init_lcd_(const uint8_t *init_cmd) {
+  ESP_LOGD(TAG, "init_lcd_");
+  uint8_t cmd, x, num_args;
+  const uint8_t *addr = init_cmd;
+
+  while ((cmd = pgm_read_byte(addr++)) > 0) {
+    x = pgm_read_byte(addr++);
```

Comment:
```suggestion
    x = progmem_read_byte(addr++);
```

That's defined in `core/hal.h` and will be cross-platform.